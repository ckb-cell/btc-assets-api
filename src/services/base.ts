import axios, { AxiosInstance } from 'axios';
import { FastifyBaseLogger } from 'fastify';

export abstract class BaseRequestService {
  protected request: AxiosInstance;

  constructor(baseURL: string, headers?: Record<string, string>) {
    this.request = axios.create({
      baseURL,
      headers,
    });
  }

  public setLogger(logger: FastifyBaseLogger) {
    const constructor = this.constructor.name;
    this.request.interceptors.request.use((config) => {
      logger.info(`[${constructor}] ${JSON.stringify(config.data)}`);
      return config;
    });

    this.request.interceptors.response.use(
      (response) => {
        logger.info(`[${constructor}] ${response.status} ${JSON.stringify(response.data)}`);
        return response;
      },
      (error) => {
        logger.error(
          `[${constructor}] ${error.response?.status} ${JSON.stringify(error.response?.data)}`,
        );
        return Promise.reject(error);
      },
    );
  }
}
