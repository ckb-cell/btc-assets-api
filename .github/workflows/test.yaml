---
name: Unit Tests

on:
  pull_request:
    branches:
      - main
      - develop

jobs:
  test:
    runs-on: ubuntu-latest

    services:
      redis:
        image: redis
        options: >-
          --health-cmd "redis-cli ping"
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 6379:6379

    steps:
      - name: Checkout Code
        uses: actions/checkout@v3

      - name: Install PNPM
        uses: pnpm/action-setup@v3
        with:
          version: 8

      - name: Use Node.js
        uses: actions/setup-node@v3
        with:
          node-version: "21"
          cache: "pnpm"

      - name: Install dependencies
        run: pnpm install

      - name: Run Unit Tests
        env:
          JWT_SECRET: ${{ secrets.JWT_SECRET }}
          BITCOIN_JSON_RPC_URL: ${{ secrets.BITCOIN_JSON_RPC_URL }}
          BITCOIN_JSON_RPC_USERNAME: ${{ secrets.BITCOIN_JSON_RPC_USERNAME }}
          BITCOIN_JSON_RPC_PASSWORD: ${{ secrets.BITCOIN_JSON_RPC_PASSWORD }}
          BITCOIN_ELECTRS_API_URL: ${{ secrets.BITCOIN_ELECTRS_API_URL }}
          BITCOIN_SPV_SERVICE_URL: ${{ secrets.BITCOIN_SPV_SERVICE_URL }}
          CKB_RPC_URL: ${{ secrets.CKB_RPC_URL }}
          CKB_INDEXER_URL: ${{ secrets.CKB_INDEXER_URL }}
          PAYMASTER_PRIVATE_KEY: ${{ secrets.PAYMASTER_PRIVATE_KEY }}
          REDIS_URL: redis://localhost:6379
          CI_REDIS_URL: redis://localhost:6379
        run: pnpm test
