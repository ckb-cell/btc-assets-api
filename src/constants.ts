export const CUSTOM_HEADERS = {
  ApiCache: 'x-api-cache',
  ResponseCacheable: 'x-response-cacheable',
};

export enum ApiCacheStatus {
  Hit = 'HIT',
  Miss = 'MISS',
}

export const JWT_IGNORE_URLS = ['/token', '/docs'];
