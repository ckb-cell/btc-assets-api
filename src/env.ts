import 'dotenv/config';
import z from 'zod';
import process from 'node:process';
import { omit } from 'lodash';

const envSchema = z.object({
  NODE_ENV: z.string().default('development'),
  PORT: z.string().optional(),
  ADDRESS: z.string().optional(),
  NETWORK: z.enum(['mainnet', 'testnet']).default('testnet'),

  DOMAIN: z.string().optional(),

  SENTRY_DSN_URL: z.string().optional(),
  REDIS_URL: z.string(),
  RATE_LIMIT_PER_MINUTE: z.number().default(100),
  LOGGER_LEVEL: z.enum(['debug', 'info', 'warn', 'error']).default('info'),

  ADMIN_USERNAME: z.string().optional(),
  ADMIN_PASSWORD: z.string().optional(),

  /**
   * JWT_SECRET is used to sign the JWT token for authentication.
   */
  JWT_SECRET: z.string(),
  /**
   * The URL/USERNAME/PASSWORD of the Bitcoin JSON-RPC server.
   * The JSON-RPC server is used to query the Bitcoin blockchain.
   */
  BITCOIN_JSON_RPC_URL: z.string(),
  BITCOIN_JSON_RPC_USERNAME: z.string(),
  BITCOIN_JSON_RPC_PASSWORD: z.string(),
  /**
   * The URL of the Electrs API.
   * Electrs is a Rust implementation of Electrum Server.
   * It is used to query the Bitcoin blockchain (balance, transactions, etc).
   */
  BITCOIN_ELECTRS_API_URL: z.string(),

  /**
   * Bitcoin SPV service URL
   * https://github.com/ckb-cell/ckb-bitcoin-spv-service
   */
  BITCOIN_SPV_SERVICE_URL: z.string(),
  /**
   * The URL of the CKB JSON-RPC server.
   */
  CKB_RPC_URL: z.string(),
  /**
   * Paymaster private key, used to sign the transaction with paymaster cell.
   */
  PAYMASTER_PRIVATE_KEY: z.string(),
  /**
   * Paymaster cell capacity in shannons
  * (254 CKB for RGB++ capacity + 61 CKB for change cell capacity + 1 CKB for fee cell)
   */
  PAYMASTER_CELL_CAPACITY: z.coerce.number().default(316 * 10 ** 8),
  PAYMASTER_CELL_PRESET_COUNT: z.coerce.number().default(500),
  PAYMASTER_CELL_REFILL_THRESHOLD: z.coerce.number().default(0.3),

  UNLOCKER_CRON_SCHEDULE: z.string().default('*/5 * * * *'),
  UNLOCKER_CELL_BATCH_SIZE: z.coerce.number().default(100),
  UNLOCKER_MONITOR_SLUG: z.string().default('btctimelock-cells-unlock'),

  TRANSACTION_QUEUE_JOB_DELAY: z.coerce.number().default(120 * 1000),
});

export type Env = z.infer<typeof envSchema>;
export const env = envSchema.parse(process.env);

export const getSafeEnvs = () =>
  omit(env, ['ADMIN_PASSWORD', 'JWT_SECRET', 'BITCOIN_JSON_RPC_PASSWORD', 'PAYMASTER_PRIVATE_KEY']);

export const isGenerateTokenPrivate = env.NODE_ENV === 'production' && env.ADMIN_USERNAME && env.ADMIN_PASSWORD;
