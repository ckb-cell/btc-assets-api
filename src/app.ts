import { FastifyInstance } from 'fastify';
import { AxiosError } from 'axios';
import bitcoinRoutes from './routes/bitcoin';
import { env } from './env';
import tokenRoutes from './routes/token';
import swaggerPlugin from './plugins/swagger';
import jwtPlugin from './plugins/jwt';
import cachePlugin from './plugins/cache';
import rateLimitPlugin from './plugins/rate-limit';
import * as Sentry from '@sentry/node';

if (env.SENTRY_DSN_URL && env.NODE_ENV !== 'development') {
  Sentry.init({
    dsn: env.SENTRY_DSN_URL,
    tracesSampleRate: 1.0,
  });
}

async function routes(fastify: FastifyInstance) {
  fastify.register(rateLimitPlugin);
  fastify.register(swaggerPlugin);
  fastify.register(jwtPlugin);
  fastify.register(cachePlugin);

  fastify.register(tokenRoutes, { prefix: '/token' });
  fastify.register(bitcoinRoutes, { prefix: '/bitcoin/v1' });

  fastify.setErrorHandler((error, _, reply) => {
    Sentry.captureException(error);
    if (error instanceof AxiosError) {
      reply
        .status(error.response?.status || 500)
        .send({ ok: false, error: error.response?.data.error ?? error.message });
      return;
    }
    reply
      .status(error.statusCode ?? 500)
      .send({ ok: false, statusCode: error.statusCode, message: error.message });
  });
}

export default routes;
