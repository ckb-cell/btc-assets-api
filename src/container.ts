import { createContainer, InjectionMode, asValue, asClass, asFunction } from 'awilix';
import { Redis } from 'ioredis';
import pino from 'pino';
import Bitcoind from './services/bitcoind';
import ElectrsAPI from './services/electrs';
import { env } from './env';
import TransactionManager from './services/transaction';
import Paymaster from './services/paymaster';
import { RPC as CkbRPC, Indexer as CkbIndexer } from '@ckb-lumos/lumos';
import Unlocker from './services/unlocker';

export interface Cradle {
  env: typeof env;
  logger: pino.BaseLogger;
  redis: Redis;
  ckbRpc: CkbRPC;
  ckbIndexer: CkbIndexer;
  bitcoind: Bitcoind;
  electrs: ElectrsAPI;
  paymaster: Paymaster;
  unlocker: Unlocker;
  transactionManager: TransactionManager;
}

const container = createContainer<Cradle>({
  injectionMode: InjectionMode.PROXY,
  strict: true,
});

container.register({
  env: asValue(env),
  logger: asValue(pino()),
  redis: asValue(
    new Redis(env.REDIS_URL, {
      maxRetriesPerRequest: null,
    }),
  ),
  ckbRpc: asFunction(() => new CkbRPC(env.CKB_RPC_URL)).singleton(),
  ckbIndexer: asFunction(() => new CkbIndexer(env.CKB_INDEXER_URL)).singleton(),
  bitcoind: asClass(Bitcoind).singleton(),
  electrs: asClass(ElectrsAPI).singleton(),
  paymaster: asClass(Paymaster).singleton(),
  transactionManager: asClass(TransactionManager).singleton(),
  unlocker: asClass(Unlocker).singleton(),
});

export default container;
