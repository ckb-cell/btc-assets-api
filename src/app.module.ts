import { Module } from '@nestjs/common';
import { AppService } from './app.service';
import { TokenModule } from './modules/token/token.module';
import { BitcoinModule } from './modules/bitcoin/bitcoin.module';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { envSchema } from './env';
import { SentryModule } from '@ntegral/nestjs-sentry';
import { nodeProfilingIntegration } from '@sentry/profiling-node';
import { CacheModule } from '@nestjs/cache-manager';
import type { RedisClientOptions } from 'redis';
import * as redisStore from 'cache-manager-redis-store';
import { StoreConfig } from 'cache-manager';

@Module({
  imports: [
    ConfigModule.forRoot({
      isGlobal: true,
      validate: envSchema.parse,
    }),
    CacheModule.registerAsync<RedisClientOptions>({
      isGlobal: true,
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) =>
        ({
          store: redisStore,
          url: configService.get('REDIS_URL'),
          ttl: 10,
        }) as StoreConfig,
    }),
    SentryModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        dsn: configService.get('SENTRY_DSN'),
        environment: configService.get('NODE_ENV'),
        tracesSampleRate: 0.5,
        profilesSampleRate: 0.5,
        integrations: [nodeProfilingIntegration()],
        logLevels:
          configService.get('NODE_ENV') === 'production'
            ? ['warn', 'error']
            : ['debug', 'log', 'warn', 'error'],
        beforeSend: (event) => {
          if (event.level === 'error') {
            return event;
          }
          return null;
        },
      }),
    }),
    TokenModule,
    BitcoinModule,
  ],
  controllers: [],
  providers: [AppService],
})
export class AppModule {}
