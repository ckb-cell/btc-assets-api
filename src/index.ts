import process from 'node:process';
import { env } from './env';
import { buildFastify } from './app';

const port = parseInt(env.PORT || '3000', 10);
const host = env.ADDRESS || '0.0.0.0';

const app = buildFastify();

app.listen({ port, host }, (err, address) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }

  // eslint-disable-next-line no-console
  console.log(`Server listening at ${address}`);
});
