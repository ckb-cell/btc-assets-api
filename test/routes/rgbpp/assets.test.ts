import { beforeEach, expect, test } from 'vitest';
import { buildFastify } from '../../../src/app';
import { describe } from 'node:test';
import { blockchain } from '@ckb-lumos/base';
import { hexify } from '@ckb-lumos/codec/lib/bytes';

let token: string;

describe('/rgbpp/v1/assets', () => {
  beforeEach(async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'POST',
      url: '/token/generate',
      payload: {
        app: 'test',
        domain: 'test.com',
      },
    });
    const data = response.json();
    token = data.token;

    await fastify.close();
  });

  test('Get RGB++ assets by BTC txid', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'GET',
      url: '/rgbpp/v1/assets/ca159e04767c25cb012f0d1c0731c767e2b58468d4cd7b505de0b184dcf97017',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });

  test('Get RGB++ assets by BTC txid and vout', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'GET',
      url: '/rgbpp/v1/assets/ca159e04767c25cb012f0d1c0731c767e2b58468d4cd7b505de0b184dcf97017/1',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });

  test('Get RGB++ assets type info (xUDT)', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const typeScript = blockchain.Script.pack({
      codeHash: '0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb',
      hashType: 'type',
      args: '0x8c556e92974a8dd8237719020a259d606359ac2cc958cb8bda77a1c3bb3cd93b',
    });
    const response = await fastify.inject({
      method: 'GET',
      url: '/rgbpp/v1/assets/type?type_script=' + hexify(typeScript),
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });

  test('Get RGB++ assets type info (Spore)', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const typeScript = blockchain.Script.pack({
      codeHash: '0x685a60219309029d01310311dba953d67029170ca4848a4ff638e57002130a0d',
      hashType: 'data1',
      args: '0xc966bd9131e7f23629a022b1310adfb3c5c98c1ac0c0ce11871685b4c8199959',
    });
    const response = await fastify.inject({
      method: 'GET',
      url: '/rgbpp/v1/assets/type?type_script=' + hexify(typeScript),
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });
});
