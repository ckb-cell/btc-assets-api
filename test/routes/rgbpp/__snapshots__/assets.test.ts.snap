// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Get RGB++ assets by BTC txid 1`] = `
[
  {
    "blockNumber": "0xc6067f",
    "cellOutput": {
      "capacity": "0x5e9f52520",
      "lock": {
        "args": "0x010000001770f9dc84b1e05d507bcdd46884b5e267c731071c0d2f01cb257c76049e15ca",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0x7e96affc517fe2b286e309e88d6b331911fc319baa3f8a830dcf59a9756f2c25",
        "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
        "hashType": "type",
      },
    },
    "data": "0x0040075af07507000000000000000000",
    "outPoint": {
      "index": "0x0",
      "txHash": "0xb799ae6896875f2a1a1e765d614cf864c444242babc4bfd7808b37fdea66c4cc",
    },
    "txIndex": "0x2",
    "typeHash": "0x27b36bd74aec3ffccb1c1351c3e3db64cba33ee1b5651059fe468fbcbb8c81bd",
  },
]
`;

exports[`Get RGB++ assets by BTC txid and vout 1`] = `
[
  {
    "blockNumber": "0xc6067f",
    "cellOutput": {
      "capacity": "0x5e9f52520",
      "lock": {
        "args": "0x010000001770f9dc84b1e05d507bcdd46884b5e267c731071c0d2f01cb257c76049e15ca",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "type": {
        "args": "0x7e96affc517fe2b286e309e88d6b331911fc319baa3f8a830dcf59a9756f2c25",
        "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
        "hashType": "type",
      },
    },
    "data": "0x0040075af07507000000000000000000",
    "outPoint": {
      "index": "0x0",
      "txHash": "0xb799ae6896875f2a1a1e765d614cf864c444242babc4bfd7808b37fdea66c4cc",
    },
    "txIndex": "0x2",
    "typeHash": "0x27b36bd74aec3ffccb1c1351c3e3db64cba33ee1b5651059fe468fbcbb8c81bd",
  },
]
`;

exports[`Get RGB++ assets type info (Spore) 1`] = `
{
  "cluster": {
    "description": "{"description":"Lower body gear is custom-made to provide comfort, mobility, and fashion for our short-legged Nervapes, gearing them up for whatever adventure lies ahead.","dob":{"ver":0,"decoder":{"type":"code_hash","hash":"0x1c84212ebd817e9de09d2a79f85cc421b684eda63409cfa75688f98716e77b5f"},"pattern":[["prev.type","string",0,1,"options",["image"]],["prev.bg","string",0,74,"utf8"],["prev.view","number",74,1,"range",[0,255]],["Entity","number",75,8,"raw"]]}}",
    "id": "0x99f2d34427694b1188d7ba462bd43e82c15f3a8238adab3bf5863c9bf97c9373",
    "name": "Nervape Lower Body",
  },
  "contentType": "dob/0",
  "type": "spore",
}
`;

exports[`Get RGB++ assets type info (xUDT) 1`] = `
{
  "decimal": 8,
  "name": "XUDT Test Token",
  "symbol": "PDD",
  "type": "xudt",
  "type_hash": "0x10f511f2efb0027191b97ac5b4bd77374ffdac7399e8527d76f5f9bd32e7d35b",
  "type_script": {
    "args": "0x8c556e92974a8dd8237719020a259d606359ac2cc958cb8bda77a1c3bb3cd93b",
    "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
    "hashType": "type",
  },
}
`;
